raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java

```diff
@@ -144,6 +204,119 @@ public void testInitializeAsCandidateAndBecomeLeader() throws Exception {
+    }
+
+    @Test
+    public void testHandleBeginQuorumResponse() throws Exception {
+        int otherNodeId = 1;
+        int leaderEpoch = 2;
+        Set<Integer> voters = Utils.mkSet(localId, otherNodeId);
+
+        MockQuorumStateStore electionStore = new MockQuorumStateStore() {
+            @Override
+            public ElectionState readElectionState() {
+                return ElectionState.withElectedLeader(leaderEpoch, localId);
+            }
+        };
+        KafkaRaftClient client = buildClient(voters, electionStore);
+        assertTrue(client.quorumState().isLeader());
+        assertEquals(Collections.singleton(otherNodeId), client.quorumState().leaderStateOrThrow().nonEndorsingFollowers());
+
+        BeginQuorumEpochResponseData beginQuorumEpochRequest = new BeginQuorumEpochResponseData()
+                                                                   .setLeaderId(localId)
+                                                                   .setLeaderEpoch(leaderEpoch);
+        channel.mockReceive(new RaftResponse.Inbound(channel.newCorrelationId(), beginQuorumEpochRequest, otherNodeId));
+
+        client.poll();
+        assertTrue(client.quorumState().isLeader());
+        assertEquals(Collections.emptySet(), client.quorumState().leaderStateOrThrow().nonEndorsingFollowers());
+    }
+
+    @Test
+    public void testHandleEndQuorumRequest() throws Exception {
+        int oldLeaderId = 1;
+        int leaderEpoch = 2;
+        Set<Integer> voters = Utils.mkSet(localId, oldLeaderId);
+
+        MockQuorumStateStore electionStore = new MockQuorumStateStore() {
+            @Override
+            public ElectionState readElectionState() {
+                return ElectionState.withElectedLeader(leaderEpoch, oldLeaderId);
+            }
+        };
+        KafkaRaftClient client = buildClient(voters, electionStore);
+        assertTrue(client.quorumState().isFollower());
+        assertTrue(client.quorumState().followerStateOrThrow().hasLeader());
+
+        EndQuorumEpochRequestData endQuorumEpochRequest = new EndQuorumEpochRequestData()
+                                                                  .setLeaderId(oldLeaderId)
+                                                                  .setLeaderEpoch(leaderEpoch);
+        channel.mockReceive(new RaftRequest.Inbound(channel.newCorrelationId(), endQuorumEpochRequest, time.milliseconds()));
+
+        pollUntilSend(client);
+
+
+        List<RaftMessage> sentMessages = channel.drainSendQueue();
+        assertEquals(2, sentMessages.size());
+        RaftMessage raftMessage = sentMessages.get(0);
+        assertTrue(raftMessage.data() instanceof FindQuorumRequestData);
+        FindQuorumRequestData request = (FindQuorumRequestData) raftMessage.data();
+        assertEquals(localId, request.replicaId());
+        int findQuorumCorrelationId = raftMessage.correlationId();
+
+        raftMessage = sentMessages.get(1);
```

Comment:
Ordering assumptions like this tend to be a little brittle over time. A more reliable approach is to build a utility which lets us pull the specific request type we are looking for. For example, maybe we could add `drainSendQueue(ApiKeys)`?

Reply:
Sg

Reply:
Was this comment addressed?

Reply:
Ah, will do.